src/module.c

```diff
@@ -10388,6 +10393,154 @@ int RM_GetDbIdFromDefragCtx(RedisModuleDefragCtx *ctx) {
     return ctx->dbid;
 }
 
+/* --------------------------------------------------------------------------
+ * ## Modules EventLoop API
+ * --------------------------------------------------------------------------*/
+
+typedef struct EventLoopData {
+    RedisModuleEventLoopCallback rProc;
+    RedisModuleEventLoopCallback wProc;
+    void *user_data;
+} EventLoopData;
+
+static int eventLoopToAeMask(int mask) {
+    int aeMask = 0;
+    if (mask & REDISMODULE_EVENTLOOP_READABLE)
+        aeMask |= AE_READABLE;
+    if (mask & REDISMODULE_EVENTLOOP_WRITABLE)
+        aeMask |= AE_WRITABLE;
+    return aeMask;
+}
+
+static int eventLoopFromAeMask(int ae_mask) {
+    int mask = 0;
+    if (ae_mask & AE_READABLE)
+        mask |= REDISMODULE_EVENTLOOP_READABLE;
+    if (ae_mask & AE_WRITABLE)
+        mask |= REDISMODULE_EVENTLOOP_WRITABLE;
+    return mask;
+}
+
+static void eventLoopCbReadable(struct aeEventLoop *ae, int fd, void *user_data, int ae_mask) {
+    UNUSED(ae);
+    EventLoopData *data = user_data;
+    data->rProc(fd, data->user_data, eventLoopFromAeMask(ae_mask));
+}
+
+static void eventLoopCbWritable(struct aeEventLoop *ae, int fd, void *user_data, int ae_mask) {
+    UNUSED(ae);
+    EventLoopData *data = user_data;
+    data->wProc(fd, data->user_data, eventLoopFromAeMask(ae_mask));
+}
+
+/* Add a pipe / socket event to the event loop
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * ERANGE: 'fd' is negative or higher than 'maxclients' Redis config.
+ * * EINVAL: 'callback' is NULL or 'mask' value is invalid.
+ *   errno might take other values in case of an internal error.
+ *
+ *  Example:
+ *  void onReadable(int fd, void *user_data, int mask) {
+ *      char buf[32];
+ *      int bytes = read(fd,buf,sizeof(buf));
+ *      printf("Read %d bytes \n", bytes);
+ *  }
+ *  RM_EventLoopAdd(fd, REDISMODULE_EVENTLOOP_READABLE, onReadable, NULL);
+ */
+int RM_EventLoopAdd(int fd, int mask, RedisModuleEventLoopCallback callback, void *user_data) {
+    if (fd < 0 || fd >= aeGetSetSize(server.el)) {
+        errno = ERANGE;
+        return REDISMODULE_ERR;
+    }
+
+    if (!callback || mask & ~(REDISMODULE_EVENTLOOP_READABLE |
+                              REDISMODULE_EVENTLOOP_WRITABLE)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    /* We are going to register stub callbacks to 'ae' for two reasons :
+     * - "ae" callback signature is different from RedisModuleEventLoopCallback,
+     *   that will be handled it in our stub callbacks.
+     * - We need to remap 'mask' value to provide binary compatibility.
+     *
+     *  For the stub callbacks, saving user 'callback' and 'user_data' in an
+     *  EventLoopData object and passing it to ae, later we'll extract
+     *  'callback' and 'user_data' from that.
```

Comment:
Not important, but without this extra space it is more clear that it's normal text, not part of a list, etc.

```suggestion
     * For the stub callbacks, saving user 'callback' and 'user_data' in an
     * EventLoopData object and passing it to ae, later we'll extract
     * 'callback' and 'user_data' from that.
```

Reply:
... and maybe use backticks for `callback`, `user_data`, etc.

Reply:
Added backticks as well. Thank you very much for the review @zuiderkwast.